apps/chat/src/components/Chat/ChatMessage/ChatMessageTemplatesModal/ChatMessageTemplatesModal.tsx

```diff
@@ -0,0 +1,244 @@
+  conversation: Conversation;
+}
+
+const EMPTY_ROW = ['', ''];
+
+export const ChatMessageTemplatesModal = ({
+  isOpen,
+  onClose,
+  message,
+  conversation,
+}: Props) => {
+  const { t } = useTranslation(Translation.Chat);
+  const dispatch = useAppDispatch();
+  const showMore = message.content.length > 160;
+  const [collapsed, setCollapsed] = useState(showMore);
+  const [previewMode, setPreviewMode] = useState(false);
+  const [templates, setTemplates] = useState([
+    ...Object.entries(message.templateMapping ?? {}),
+    EMPTY_ROW,
+  ]);
+
+  useEffect(() => {
+    if (isOpen) {
+      setTemplates([
+        ...Object.entries(message.templateMapping ?? {}),
+        EMPTY_ROW,
+      ]);
+    }
+  }, [message.templateMapping, isOpen]);
+
+  const handleChangeTemplate = useCallback(
+    (index: number, content: string, template: string) => {
+      const newTemplates = [...templates];
+      newTemplates[index] = [content, template];
+      if (index === newTemplates.length - 1 && (content || template)) {
+        newTemplates.push(EMPTY_ROW);
+      }
+      setTemplates(newTemplates);
+    },
+    [templates],
+  );
+
+  const handleDeleteTemplate = useCallback(
+    (index: number) => {
+      const newTemplates = [...templates];
+      newTemplates.splice(index, 1);
+      setTemplates(newTemplates);
+    },
+    [templates],
+  );
+
+  const handleSaveTemplate = useCallback(() => {
+    const templateMapping = Object.fromEntries(
+      templates.slice(0, templates.length - 1),
+    );
+    const messages = conversation.messages.map((mes) =>
+      mes === message ? { ...mes, templateMapping } : mes,
+    );
+    dispatch(
+      ConversationsActions.updateConversation({
+        id: conversation.id,
+        values: {
+          messages,
+        },
+      }),
+    );
+    onClose(true);
+  }, [
+    conversation.id,
+    conversation.messages,
+    dispatch,
+    message,
+    onClose,
+    templates,
+  ]);
+
+  const templateResult = useMemo(() => {
+    return templates.reduce(
+      (acc, [key, value]) => acc.replaceAll(key, value),
+      message.content,
+    );
+  }, [message.content, templates]);
+
+  const isInvalid = useMemo(
+    () =>
+      templates
+        .slice(0, templates.length - 1)
+        .some(
+          ([content, template]) =>
+            !content ||
+            !template ||
+            message.content.indexOf(content) === -1 ||
+            !PROMPT_VARIABLE_REGEX.test(template) ||
+            !templateMatchContent(content, template),
+        ),
+    [message.content, templates],
+  );
+
+  return (
+    <Modal
+      portalId="theme-main"
+      state={isOpen ? ModalState.OPENED : ModalState.CLOSED}
+      onClose={() => onClose(false)}
+      dataQa="message-templates-dialog"
+      containerClassName="h-fit max-h-full inline-block w-full min-w-[90%] text-center md:min-w-[300px] md:max-w-[880px] flex flex-col"
+      dismissProps={{ outsidePressEvent: 'mousedown', outsidePress: true }}
```

Comment:
move it out of component? To prevent unnecessary rerenders. Maybe we need to check it everywhere someday